In [1]:
import pandas as pd
import numpy as np
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split, GridSearchCV

In [2]:
data=pd.read_csv('College.csv')
df=pd.DataFrame(data)
df.isna().sum()

Private        0
Apps           0
Accept         0
Enroll         0
Top10perc      0
Top25perc      0
F.Undergrad    0
P.Undergrad    0
Outstate       0
Room.Board     0
Books          0
Personal       0
PhD            0
Terminal       0
S.F.Ratio      0
perc.alumni    0
Expend         0
Grad.Rate      0
dtype: int64

In [3]:
l=LabelEncoder()
df['LabelPrivate']=l.fit_transform(df['Private'])
df.drop('Private', axis=1, inplace=True)

In [4]:
df.head(7)

Apps  Accept  Enroll  Top10perc  Top25perc  F.Undergrad  P.Undergrad  \
0  1660    1232     721         23         52         2885          537   
1  2186    1924     512         16         29         2683         1227   
2  1428    1097     336         22         50         1036           99   
3   417     349     137         60         89          510           63   
4   193     146      55         16         44          249          869   
5   587     479     158         38         62          678           41   
6   353     340     103         17         45          416          230   

   Outstate  Room.Board  Books  Personal  PhD  Terminal  S.F.Ratio  \
0      7440        3300    450      2200   70        78       18.1   
1     12280        6450    750      1500   29        30       12.2   
2     11250        3750    400      1165   53        66       12.9   
3     12960        5450    450       875   92        97        7.7   
4      7560        4120    800      1500   76        72       11.9   
5     13500        3335    500       675   67        73        9.4   
6     13290        5720    500      1500   90        93       11.5   

   perc.alumni  Expend  Grad.Rate  LabelPrivate  
0           12    7041         60             1  
1           16   10527         56             1  
2           30    8735         54             1  
3           37   19016         59             1  
4            2   10922         15             1  
5           11    9727         55             1  
6           26    8861         63             1

In [5]:
df.columns

Index(['Apps', 'Accept', 'Enroll', 'Top10perc', 'Top25perc', 'F.Undergrad',
       'P.Undergrad', 'Outstate', 'Room.Board', 'Books', 'Personal', 'PhD',
       'Terminal', 'S.F.Ratio', 'perc.alumni', 'Expend', 'Grad.Rate',
       'LabelPrivate'],
      dtype='object')

In [6]:
features = ['Apps', 'Accept', 'Enroll', 'Top10perc', 'Top25perc', 'F.Undergrad',
             'P.Undergrad', 'Outstate', 'Room.Board', 'Books', 'Personal', 'PhD',
            'Terminal', 'S.F.Ratio', 'perc.alumni', 'Expend', 'Grad.Rate']
label=['LabelPrivate']
X=df[features]
Y=df[label]
seed=1234
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=.20,random_state=seed)

####    3.Fit a linear svm from scikit learn and observe the accuracy

In [7]:
svc=SVC(kernel='linear', random_state=seed)
model=svc.fit(X_train,Y_train)
predvalue=model.predict(X_test)
accuracyscore = accuracy_score(Y_test,predvalue)
print("SVC Initial Accuracy score = ",accuracyscore)
# predvalue=model.predict(X_train)
# accuracyscore = accuracy_score(Y_train,predvalue)
# print("SVC Train Overfitting = ",accuracyscore)

C:\Users\bpeddi\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVC Initial Accuracy score =  0.9423076923076923


#### 4.Preprocess the data using StandardScalar - Accuracy is same 94.23%

In [8]:
ss = StandardScaler()
scaled_data=ss.fit_transform(X)
X_scaled = pd.DataFrame(scaled_data, columns=X.columns)

In [9]:
X_scaled.head(3)

Apps    Accept    Enroll  Top10perc  Top25perc  F.Undergrad  \
0 -0.346882 -0.321205 -0.063509  -0.258583  -0.191827    -0.168116   
1 -0.210884 -0.038703 -0.288584  -0.655656  -1.353911    -0.209788   
2 -0.406866 -0.376318 -0.478121  -0.315307  -0.292878    -0.549565   

   P.Undergrad  Outstate  Room.Board     Books  Personal       PhD  Terminal  \
0    -0.209207 -0.746356   -0.964905 -0.602312  1.270045 -0.163028 -0.115729   
1     0.244307  0.457496    1.909208  1.215880  0.235515 -2.675646 -3.378176   
2    -0.497090  0.201305   -0.554317 -0.905344 -0.259582 -1.204845 -0.931341   

   S.F.Ratio  perc.alumni   Expend  Grad.Rate  
0   1.013776    -0.867574 -0.50191  -0.318252  
1  -0.477704    -0.544572  0.16611  -0.551262  
2  -0.300749     0.585935 -0.17729  -0.667767

In [10]:
X.head(3)

Apps  Accept  Enroll  Top10perc  Top25perc  F.Undergrad  P.Undergrad  \
0  1660    1232     721         23         52         2885          537   
1  2186    1924     512         16         29         2683         1227   
2  1428    1097     336         22         50         1036           99   

   Outstate  Room.Board  Books  Personal  PhD  Terminal  S.F.Ratio  \
0      7440        3300    450      2200   70        78       18.1   
1     12280        6450    750      1500   29        30       12.2   
2     11250        3750    400      1165   53        66       12.9   

   perc.alumni  Expend  Grad.Rate  
0           12    7041         60  
1           16   10527         56  
2           30    8735         54

In [11]:
X_train,X_test,Y_train,Y_test=train_test_split(X_scaled,Y,test_size=.20,random_state=seed)

In [12]:
svc=SVC(kernel='linear', random_state=seed)
model=svc.fit(X_train,Y_train)
predvalue=model.predict(X_test)
accuracyscore = accuracy_score(Y_test,predvalue)
print("SVC scaled Accuracy score = ",accuracyscore)

SVC scaled Accuracy score =  0.9423076923076923


C:\Users\bpeddi\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


####  5.Use scikit learn’s gridsearch to select the best hyperparameter 

In [19]:
parameters = {'kernel':['rbf'],
              'gamma': [1e-3, 1e-4],
              'C': [1, 10, 100, 1000]}
svc2 = SVC()
clf=GridSearchCV(svc2, parameters)
clf.fit(X_train, Y_train)

C:\Users\bpeddi\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:1978: FutureWarning: The default value of cv will change from 3 to 5 in version 0.22. Specify it explicitly to silence this warning.
  warnings.warn(CV_WARNING, FutureWarning)
C:\Users\bpeddi\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\bpeddi\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\bpeddi\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y 

GridSearchCV(cv='warn', error_score='raise-deprecating',
             estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='auto_deprecated', kernel='rbf', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             iid='warn', n_jobs=None,
             param_grid={'C': [1, 10, 100, 1000], 'gamma': [0.001, 0.0001],
                         'kernel': ['rbf']},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

In [20]:
clf.cv_results_

{'mean_fit_time': array([0.01036119, 0.0125347 , 0.00902422, 0.01404087, 0.01470788,
        0.00936413, 0.01136398, 0.00868948]),
 'std_fit_time': array([0.00047188, 0.00354837, 0.00081848, 0.00245905, 0.00519716,
        0.00094476, 0.00189105, 0.00170616]),
 'mean_score_time': array([0.00401171, 0.00417988, 0.00367634, 0.00568493, 0.00534638,
        0.00367419, 0.00167155, 0.00217493]),
 'std_score_time': array([1.63258223e-06, 2.39787648e-04, 4.72494310e-04, 1.70681321e-03,
        2.87554038e-03, 4.70812420e-04, 4.72720055e-04, 2.37484719e-04]),
 'param_C': masked_array(data=[1, 1, 10, 10, 100, 100, 1000, 1000],
              mask=[False, False, False, False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'param_gamma': masked_array(data=[0.001, 0.0001, 0.001, 0.0001, 0.001, 0.0001, 0.001,
                    0.0001],
              mask=[False, False, False, False, False, False, False, False],
        fill_value='?',
             dtype=object),


In [21]:
clf.best_params_

{'C': 1000, 'gamma': 0.001, 'kernel': 'rbf'}

In [22]:
clf.best_estimator_

SVC(C=1000, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma=0.001, kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [23]:
best_model=clf.best_estimator_
pred_value = best_model.predict(X_test)
accuracyscore = accuracy_score(Y_test,pred_value)
print("SVC Grid Search Accuracy score = ",accuracyscore)

SVC Grid Search Accuracy score =  0.9551282051282052
